Code/EnginePlugins/OpenXRPlugin/OpenXRSingleton.cpp

```diff
@@ -0,0 +1,1122 @@
+  m_layer.viewCount = 2;
+  m_layer.views = m_projectionLayerViews;
+
+  ezHybridArray<XrCompositionLayerBaseHeader*, 1> layers;
+  layers.PushBack(reinterpret_cast<XrCompositionLayerBaseHeader*>(&m_layer));
+
+  // Submit the composition layers for the predicted display time.
+  XrFrameEndInfo frameEndInfo{ XR_TYPE_FRAME_END_INFO };
+  frameEndInfo.displayTime = m_frameState.predictedDisplayTime;
+  frameEndInfo.environmentBlendMode = m_blendMode;
+  frameEndInfo.layerCount = layers.GetCapacity();
+  frameEndInfo.layers = layers.GetData();
+
+  EZ_PROFILE_SCOPE("xrEndFrame");
+  XR_LOG(xrEndFrame(m_session, &frameEndInfo));
+
+  return m_hColorRT;
+}
+
+void ezOpenXR::GALDeviceEventHandler(const ezGALDeviceEvent& e)
+{
+  if (e.m_Type == ezGALDeviceEvent::Type::BeforeBeginFrame)
+  {
+    BeforeBeginFrame();
+  }
+}
+
+void ezOpenXR::GameApplicationEventHandler(const ezGameApplicationExecutionEvent& e)
+{
+  EZ_ASSERT_DEV(IsInitialized(), "Need to call 'Initialize' first.");
+
+  if (e.m_Type == ezGameApplicationExecutionEvent::Type::BeforeUpdatePlugins)
+  {
+    BeforeUpdatePlugins();
+  }
+}
+
+void ezOpenXR::SetStageSpace(ezXRStageSpace::Enum space)
+{
+  m_StageSpace = space;
+}
+
+void ezOpenXR::SetHMDCamera(ezCamera* pCamera)
+{
+  EZ_ASSERT_DEV(IsInitialized(), "Need to call 'Initialize' first.");
+
+  if (m_pCameraToSynchronize == pCamera)
+    return;
+
+  m_pCameraToSynchronize = pCamera;
+  if (m_pCameraToSynchronize)
+  {
+    m_uiSettingsModificationCounter = m_pCameraToSynchronize->GetSettingsModificationCounter() + 1;
+    m_pCameraToSynchronize->SetCameraMode(ezCameraMode::Stereo, m_pCameraToSynchronize->GetFovOrDim(),
+      m_pCameraToSynchronize->GetNearPlane(), m_pCameraToSynchronize->GetFarPlane());
+  }
+
+  if (ezSoundInterface* pSoundInterface = ezSingletonRegistry::GetSingletonInstance<ezSoundInterface>())
+  {
+    pSoundInterface->SetListenerOverrideMode(m_pCameraToSynchronize != nullptr);
+  }
+}
+
+XrPosef ezOpenXR::ConvertTransform(const ezTransform& tr)
+{
+  XrPosef pose;
+  pose.orientation = ConvertOrientation(tr.m_qRotation);
+  pose.position = ConvertPosition(tr.m_vPosition);
+  return pose;
+}
+
+XrQuaternionf ezOpenXR::ConvertOrientation(const ezQuat& q)
+{
+  return {q.v.y, q.v.z, -q.v.x, -q.w};
+}
+
+XrVector3f ezOpenXR::ConvertPosition(const ezVec3& pos)
+{
+  return {pos.y, pos.z, -pos.x};
+}
+
+ezQuat ezOpenXR::ConvertOrientation(const XrQuaternionf& q)
+{
+  return {-q.z, q.x, q.y, -q.w};
+}
+
+ezVec3 ezOpenXR::ConvertPosition(const XrVector3f& pos)
+{
+  return {-pos.z, pos.x, pos.y};
+}
+
+ezMat4 ezOpenXR::ConvertPoseToMatrix(const XrPosef& pose)
+{
+  ezMat4 m;
+  ezMat3 rot = ConvertOrientation(pose.orientation).GetAsMat3();
+  ezVec3 pos = ConvertPosition(pose.position);
+  m.SetTransformationMatrix(rot, pos);
+  return m;
+}
+
+ezGALResourceFormat::Enum ezOpenXR::ConvertTextureFormat(int64_t format)
+{
+  switch (format)
```

Comment:
We should remove ezGALResourceFormat at some point and just use ezImageFormat everywhere instead.
(@marcappelsmeier)

You can use ezImageFormatMappings to map between DXGI and ezImageFormat.

You can then use ezTextureUtils to map from ezImageFormat to ezGALResourceFormat.

Reply:
I'll use that now for the color formats but I still need the switch case for the depth format as they are not in the image format enum.

Reply:
Okay, maybe it would be best to extend that then? I guess it's not in there, because a DDS file can't use the format. Again, not high priority, but we seem to have a proliferation of custom image format handling code.